### PR TITLE
grad(entr)(0.0): return inf instead of NaN

### LINKED ISSUE #15709: forward mode differentiation issue with scipy.special.xlogy
### Description

A few days ago, a [commit](https://github.com/google/jax/commit/dd023e266e6616494cdd590959103ddc646109c4) to jax.scipy.special seems to be causing my previously-working code to bomb out with NaN errors.  I believe the root problem is that you've added a custom jvp to xlogy and xlog1py which need the same safety features as the original functions.

```@custom_derivatives.custom_jvp
@_wraps(osp_special.xlogy, module='scipy.special')
def xlogy(x: ArrayLike, y: ArrayLike) -> Array:
  # Note: xlogy(0, 0) should return 0 according to the function documentation.
  x, y = promote_args_inexact("xlogy", x, y)
  x_ok = x != 0.
  safe_x = jnp.where(x_ok, x, 1.)
  safe_y = jnp.where(x_ok, y, 1.)
  return jnp.where(x_ok, lax.mul(safe_x, lax.log(safe_y)), jnp.zeros_like(x))

def _xlogy_jvp(primals, tangents):
  (x, y) = primals
  (x_dot, y_dot) = tangents
  result = xlogy(x, y)
  return result, (x_dot * lax.log(y) + y_dot * x / y).astype(result.dtype)
xlogy.defjvp(_xlogy_jvp)
```
My code is obtaining NaNs in _xlogy_jvp when I call scipy.special.entr(p) which then calls xlogy(p,p) and one of the elements of p is zero.

### What jax/jaxlib version are you using?

jax v0.4.9

### Which accelerator(s) are you using?

TPU

### Additional system info

Colab

### NVIDIA GPU info

_No response_

### PR DESCRIPTION
**Pull request by jakevdp**
Fixes #15709

As discussed with @mattjj, getting the correct answer requires evaluating the gradient of $x\log(x)$ correctly; I opted to do this by adding a new private fuinction for computing `xlogx` with the correct gradient.

### CODE REVIEW

**Review by mattjj — APPROVED** [review]
Can you add a call to `jtu.check_grads` in the test? Any time we have a `custom_jvp` I'd like to test it with that. Can do it away from 0.
